api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/handler/StringHandler.java

```diff
@@ -263,6 +277,48 @@ public void addStringData(
         observable.subscribe(new ResultSetObserver(asyncResponse));
     }
 
+    @POST
+    @Path("/raw/query")
+    @ApiOperation(value = "Fetch raw data points for multiple metrics")
+    public Response findRawData(QueryRequest query) {
+        TimeRange timeRange = new TimeRange(query.getStart(), query.getEnd());
+        if (!timeRange.isValid()) {
+            return badRequest(new ApiError(timeRange.getProblem()));
+        }
+
+        int limit;
+        if (query.getLimit() == null) {
+            limit = 0;
+        } else {
+            limit = query.getLimit();
+        }
+        Order order;
+        if (query.getOrder() == null) {
+            order = Order.defaultValue(limit, timeRange.getStart(), timeRange.getEnd());
+        } else {
+            order = Order.fromText(query.getOrder());
+        }
+
+        List<MetricId<String>> metricIds = query.getIds().stream().map(id -> new MetricId<>(tenantId, STRING, id))
+                .collect(toList());
+        Observable<NamedDataPoint<String>> dataPoints = metricsService.findDataPoints(metricIds, timeRange.getStart(),
+                timeRange.getEnd(), limit, order).observeOn(Schedulers.io());
+
+        StreamingOutput stream = output -> {
+            JsonGenerator generator = mapper.getFactory().createGenerator(output, JsonEncoding.UTF8);
+            CountDownLatch latch = new CountDownLatch(1);
+            dataPoints.subscribe(new NamedDataPointObserver<>(generator, latch, STRING));
+
+            try {
+                latch.await();
```

Comment:
Le 20/06/2016 à 23:35, jsanda a écrit :
> In
> api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/handler/StringHandler.java
> <https://github.com/hawkular/hawkular-metrics/pull/518#discussion_r67771082>:
>
>> +        } else {
>> +            order = Order.fromText(query.getOrder());
>> +        }
>> +
>> +        List<MetricId<String>> metricIds = query.getIds().stream().map(id -> new MetricId<>(tenantId, STRING, id))
>> +                .collect(toList());
>> +        Observable<NamedDataPoint<String>> dataPoints = metricsService.findDataPoints(metricIds, timeRange.getStart(),
>> +                timeRange.getEnd(), limit, order).observeOn(Schedulers.io());
>> +
>> +        StreamingOutput stream = output -> {
>> +            JsonGenerator generator = mapper.getFactory().createGenerator(output, JsonEncoding.UTF8);
>> +            CountDownLatch latch = new CountDownLatch(1);
>> +            dataPoints.subscribe(new NamedDataPointObserver<>(generator, latch, STRING));
>> +
>> +            try {
>> +                latch.await();
>
> I disagree that it goes against our reactive architecture because we do
> not have a fully async/non-blocking architecture. Undertow may be async,
> but JAX-RS certainly is not. In every endpoint in which we use
> AsyncResponse we block that thread until the response is fully realized

This is not true. The purpose of AsyncReponse is precisely to decouple 
response creation from endpoint invocation. More precisely, in our case, 
it is a thread from the computation scheduler which resumes the 
AsyncResponse when data is loaded.

> in memory. It simply is not fully reactive. With these new endpoints, we
> can be dealing with arbitrarily large result sets that could easily
> introduce lots of heap pressure, making the server not so scalable. I
> think that avoiding that memory pressure is more important than
> deferring the work to another thread which will still block until the
> response is fully realized.
>

I agree, we need to find to a solution for large responses. But 
regardless of the motivation, blocking a container task thread from the 
moment the request is received until the response is fully sent is one 
of the things we have tried to avoid.

> I tried to make the streaming work with AsyncResponse but was
> unsuccessful. If there is a way to make it work, then I am all for it.
> If we were running on a fully async stack, then what I have done with
> the CountDownLatch would be a big red flag.
>

I don't have the solution at hand. If going back to blocking handler 
implementation is the price to pay to avoid OOMEs with the current 
software stack, then fine.

But since it's a major departure from the current architecture goals, I 
believe it's worth the notice.